icu4c/source/common/uresbund.cpp

```diff
@@ -1928,24 +1992,32 @@ ures_getByKeyWithFallback(const UResourceBundle *resB,
 
     int32_t type = RES_GET_TYPE(resB->fRes);
     if(URES_IS_TABLE(type)) {
+        const char* origResPath = resB->fResPath;
+        int32_t origResPathLen = resB->fResPathLen;
         res = getTableItemByKeyPath(&resB->getResData(), resB->fRes, inKey);
         const char* key = inKey;
+        bool didRootOnce = false;
         if(res == RES_BOGUS) {
             UResourceDataEntry *dataEntry = resB->fData;
             CharString path;
             char *myPath = NULL;
             const char* resPath = resB->fResPath;
             int32_t len = resB->fResPathLen;
-            while(res == RES_BOGUS && dataEntry->fParent != NULL) { /* Otherwise, we'll look in parents */
-                dataEntry = dataEntry->fParent;
+            while(res == RES_BOGUS && (dataEntry->fParent != NULL || !didRootOnce)) { /* Otherwise, we'll look in parents */
+                if (dataEntry->fParent != NULL) {
+                    dataEntry = dataEntry->fParent;
+                } else {
+                    // We can't just stop when we get to a bundle whose fParent is NULL.  That'll work most of the time,
+                    // but if the bundle that the caller passed to us was "root" (which happens in getAllItemsWithFallback(),
+                    // this function will drop right out without doing anything if "root" doesn't contain the exact key path
+                    // specified.  In that case, we need one extra time through this loop to make sure we follow any
+                    // applicable aliases at the root level.
```

Comment:
I believe you, but do you have a concrete example at hand that this change here fixes? getAllItemsWithFallback() starting from root with which path? Just trying to wrap my head around this...

Reply:
I don't remember precisely off the top of my head, but yes, you definitely have unit-test failures if you take out this logic.  As I remember it, the situation this addressed was where you ask for `unitsNarrow/area/square-foot` in `en_GB` and it falls back to `root`, which doesn't have `unitsNarrow/area/square-foot`; it only has `unitsNarrow`, which is an alias to `unitsShort`.

As I recall, the problem was that when we get to `root`, the `dataEntry->fParent != NULL` check in line 2006 causes us to bypass the whole loop, and it's the code in that loop (particularly ll. 2029-2039) that actually detects and follows the alias, so we don't follow the alias.

Faking an extra iteration through the loop to make sure we follow the alias is cheesy, especially since it forces us through the loop an extra time when we _don't_ need to do it, but I couldn't think of a cleaner way to do it without a lot more surgery to this function than I was comfortable making.  If you can recommend a better way to do this, I'm all ears.